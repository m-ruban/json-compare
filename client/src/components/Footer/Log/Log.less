@import "~styles/variables.less";

.log {
  margin-left: auto;
  color: @secondary-color;

  &__item {
    height: 100%;
    border-radius: 0.25rem;
    font-size: 12px;
    font-weight: 500;
    text-align: right;
  }

  &__type {
    text-transform: capitalize;
    text-decoration: underline;
  }

  &__error {
    color: red;
  }
}
