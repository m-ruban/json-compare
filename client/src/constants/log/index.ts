export const LOG_INFO = 'info';
export const LOG_ERROR = 'error';
