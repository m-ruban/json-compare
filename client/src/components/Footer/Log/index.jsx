import React from 'react';
import PropTypes from 'prop-types';
import withConsumer from 'hoc/withConsumer';
import 'components/Footer/Log/Log.less';

const Log = ({ log }) => {
    if (!log.type) {
        return null;
    }

    return (
        <div className="log pr-3 py-3">
            <div className="log__item">
                <div className={`log__${log.type}`}>
                    <span className="log__type">log.{log.type}</span>: {log.text}
                </div>
            </div>
        </div>
    );
};

Log.propTypes = {
    log: PropTypes.shape({
        type: PropTypes.string,
        text: PropTypes.text,
    }),
};

export default withConsumer(Log);
