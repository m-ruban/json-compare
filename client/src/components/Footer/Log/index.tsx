import { ILog } from 'constants/interfaces';
import React from 'react';

import withConsumer from 'hoc/withConsumer';

import 'components/Footer/Log/Log.less';

const Log = ({ log }: { log: ILog }) => {
    if (!log.type) {
        return null;
    }
    return (
        <div className="log pr-3 py-3">
            <div className="log__item">
                <div className={`log__${log.type}`}>
                    <span className="log__type">log.{log.type}</span>: {log.text}
                </div>
            </div>
        </div>
    );
};

export default withConsumer(Log);
