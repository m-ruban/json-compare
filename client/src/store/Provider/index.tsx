import * as React from 'react';

import { COMPARE_REQ } from 'constants/compare';
import { EQUALITY_TYPE_STRICT } from 'constants/equality';
import { AUTHOR_LINK, FEEDBACK_LINK } from 'constants/links';
import { LOG_INFO } from 'constants/log';
import { Provider } from 'store/Context';

import { equalAlerts } from 'modules/compare';
import { parseAndLog } from 'modules/parse';

interface IProviderProps {
    children: any;
}

interface IProviderState {
    alerts: Map<string, object>;
    author: string;
    changeEqualityTypes: (event: React.SyntheticEvent) => void;
    changeObj: (event: React.SyntheticEvent) => void;
    changeTool: (value: string) => void;
    equalityTypes: string;
    feedback: string;
    isCompare: boolean;
    log: object;
    objObject: object;
    objStr: string;
    toggleCompare: (event: React.SyntheticEvent) => void;
    toolObject: object;
    toolStr: string;
}

class AppProvider extends React.Component<IProviderProps, IProviderState> {
    constructor(props) {
        super(props);
        this.state = {
            alerts: new Map(),
            author: AUTHOR_LINK,
            changeEqualityTypes: this.changeEqualityTypes.bind(this),
            changeObj: this.changeObj.bind(this),
            changeTool: this.changeTool.bind(this),
            equalityTypes: EQUALITY_TYPE_STRICT,
            feedback: FEEDBACK_LINK,
            isCompare: false,
            log: { text: '', type: null },
            objObject: {},
            objStr: '',
            toggleCompare: this.toggleCompare.bind(this),
            toolObject: {},
            toolStr: '',
        };
    }

    public toggleCompare() {
        const log = {
            text: 'toggle compare',
            type: LOG_INFO,
        };

        this.setState({ isCompare: !this.state.isCompare, log }, this.changeAlerts);
    }

    public changeEqualityTypes(event) {
        const { value } = event.target;
        const log = {
            text: 'change equality types',
            type: LOG_INFO,
        };
        this.setState({ equalityTypes: value, log }, this.changeAlerts);
    }

    public changeTool(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ toolObject: obj, toolStr: value, log }, this.changeAlerts);
    }

    public changeObj(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ objObject: obj, objStr: value, log }, this.changeAlerts);
    }

    public changeAlerts() {
        const { toolObject, objObject, equalityTypes, isCompare } = this.state;
        const isEmpty = (obj) => Object.keys(obj).length === 0;

        if (!isEmpty(toolObject) && !isEmpty(objObject) && isCompare) {
            const alerts = equalAlerts(toolObject, objObject, equalityTypes);
            let req = 0;
            let diff = 0;
            let text = '';

            alerts.forEach((alert) => {
                if (!alert.scalar) { return; }
                if (alert.res === COMPARE_REQ) {
                    req++;
                } else {
                    diff++;
                }
            });

            if (diff > 0 || req > 0) {
                text = `${diff} alert${diff > 1 ? 's' : ''}, ${req} required`;
            } else {
                text = 'objects are equal';
            }

            this.setState({
                alerts,
                log: { text, type: LOG_INFO },
            });
        }
    }

    public render() {
        return <Provider value={this.state}>{this.props.children}</Provider>;
    }
}

export default AppProvider;
