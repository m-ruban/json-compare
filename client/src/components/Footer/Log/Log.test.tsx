import React from 'react';

import Log from 'components/Footer/Log';
import { render } from 'enzyme';
import { Provider } from 'store/Context';
import { emptyLog, exampleLog } from 'tests/mocks/log';

describe('<Log />', () => {
    it('Renders correctly (empty log)', () => {
        const output = render(
            <Provider value={emptyLog}>
                <Log />
            </Provider>,
        );
        expect(output).toMatchSnapshot();
    });
    it('Renders correctly (not empty log)', () => {
        const output = render(
            <Provider value={exampleLog}>
                <Log />
            </Provider>,
        );
        expect(output).toMatchSnapshot();
    });
});
