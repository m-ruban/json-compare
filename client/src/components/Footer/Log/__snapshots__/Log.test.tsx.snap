// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Log /> Renders correctly (empty log) 1`] = `null`;

exports[`<Log /> Renders correctly (not empty log) 1`] = `
<div
  class="log pr-3 py-3"
>
  <div
    class="log__item"
  >
    <div
      class="log__info"
    >
      <span
        class="log__type"
      >
        log.info
      </span>
      : test
    </div>
  </div>
</div>
`;
